server/protocols/webadmin/webadmin-mailbox-deleted-message-vault/src/main/java/org/apache/james/webadmin/vault/routes/query/QueryElementDeserializer.java

```diff
@@ -0,0 +1,87 @@
+
+import com.fasterxml.jackson.core.JsonParser;
+import com.fasterxml.jackson.core.JsonProcessingException;
+import com.fasterxml.jackson.databind.DeserializationContext;
+import com.fasterxml.jackson.databind.JsonNode;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.fasterxml.jackson.databind.deser.std.StdDeserializer;
+import com.fasterxml.jackson.databind.exc.InvalidNullException;
+import com.fasterxml.jackson.databind.node.ObjectNode;
+
+class QueryElementDeserializer extends StdDeserializer<QueryElement> {
+    protected QueryElementDeserializer() {
+        super(QueryElement.class);
+    }
+
+    @Override
+    public QueryElement deserialize(JsonParser jsonParser, DeserializationContext deserializationContext) throws IOException, JsonProcessingException {
+        ObjectMapper mapper = (ObjectMapper) jsonParser.getCodec();
+        ObjectNode queryElement = mapper.readTree(jsonParser);
+
+        if (isQueryElement(queryElement)) {
+            return generateQuery(jsonParser, mapper, queryElement);
+        }
+        return generateCriterion(jsonParser, queryElement);
+    }
+
+    private QueryElement generateQuery(JsonParser jsonParser, ObjectMapper mapper, ObjectNode queryElement) throws IOException {
+        String combinator = Optional.ofNullable(queryElement.findValue("combinator"))
+            .map(JsonNode::textValue)
+            .orElseThrow(() -> InvalidNullException.from(jsonParser, String.class, "combinator property of the query is missing"));
+
+        String criteriaString = Optional.ofNullable(queryElement.findValue("criteria"))
+            .map(JsonNode::toString)
+            .orElseThrow(() -> InvalidNullException.from(jsonParser, List.class, "criteria property of the query is missing"));
+        List<QueryElement> criteria = mapper.readValue(criteriaString, mapper.getTypeFactory().constructCollectionType(List.class, QueryElement.class));
+
+        return new QueryDTO(combinator, criteria);
+    }
+
+    private QueryElement generateCriterion(JsonParser jsonParser, ObjectNode queryElement) throws JsonProcessingException {
+        String fieldName = Optional.ofNullable(queryElement.findValue("fieldName"))
+            .map(JsonNode::textValue)
+            .orElseThrow(() -> InvalidNullException.from(jsonParser, String.class, "fieldName property of the criterion is missing"));
+
+        String operator = Optional.ofNullable(queryElement.findValue("operator"))
+            .map(JsonNode::textValue)
+            .orElseThrow(() -> InvalidNullException.from(jsonParser, String.class, "operator property of the criterion is missing"));
+
+        String value = Optional.ofNullable(queryElement.findValue("value"))
+            .map(JsonNode::textValue)
+            .orElseThrow(() -> InvalidNullException.from(jsonParser, String.class, "value property of the criterion is missing"));
+
+        return new CriterionDTO(fieldName, operator, value);
```

Comment:
Does this mean that we steal the job of jackson?

We should pass jackson the information it needs to do the deserialization job, not do it .

Reply:
If I delegate to jackson to deserialize an object it self. I will face to the Stackoverflow exception. which call deserialize() endlessly.

So I create objects by myself and delegate children creation to jackson

Reply:
Found it, just add one more annotation, I wasn't read the tutorial carefully!
```
    @JsonDeserialize(as = CriterionDTO.class)
    class CriterionDTO implements QueryElement {}
``